Mathlib/GroupTheory/IndexNormal.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2025 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+import Mathlib.Data.Fintype.Perm
+import Mathlib.Data.Nat.Prime.Factorial
+import Mathlib.GroupTheory.Index
+
+/-! # Subgroups of small index are normal
+
+* `Subgroup.normal_of_index_eq_smallest_prime_factor`: in a finite group `G`,
+a subgroup of index equal to the smallest prime factor of `Nat.card G` is normal.
+
+* `Subgroup.normal_of_index_two`: in a group `G`, a subgroup of index 2 is normal
+(This does not require `G` to be finite.)
+
+-/
+
+variable {G : Type*} [Group G] {H : Subgroup G} {p : ℕ}
+
+namespace Subgroup
+
+open MulAction MonoidHom Nat
+
+private theorem dvd_prime {r p : ℕ} (hp : p.Prime) (h : r ∣ p !)
+    (hr : ∀ {l : ℕ} (_ : l.Prime) (_ : l ∣ r), p ≤ l) : r ∣ p := by
+  rw [← Coprime.dvd_mul_right (n := (p-1)!) _]
+  · rw [mul_factorial_pred hp.pos]; exact h
+  · rw [coprime_iff_gcd_eq_one]
+    by_contra h
+    obtain ⟨l, hl, hl'⟩ := exists_prime_and_dvd h
+    rw [dvd_gcd_iff, hl.dvd_factorial] at hl'
+    apply (lt_iff_not_ge p.pred p).mp (Nat.pred_lt hp.ne_zero)
+    rw [pred_eq_sub_one, ge_iff_le]
+    exact le_trans (hr hl hl'.left) hl'.right
+
+/-- A subgroup of a finite group whose index is the smallest prime factor is normal -/
+theorem normal_of_index_eq_smallest_prime_factor [Finite G]
+    (hp : p.Prime) (hHp : H.index = p) (hp' : ∀ {l : ℕ} (_ : l.Prime) (_: l ∣ Nat.card G), p ≤ l) :
+    H.Normal := by
```

Comment:
I think using `minFac` gives a slightly nicer statement (for instance, see [IsCyclic.isComplement'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/Transfer.html#IsCyclic.isComplement')):
```suggestion
    (hHp : H.index = (Nat.card G).minFac) :
    H.Normal := by
  by_cases hG : Nat.card G = 1
  · rw [hG, minFac_one, index_eq_one] at hHp
    rw [hHp]
    infer_instance
  have hp : (Nat.card G).minFac.Prime := minFac_prime hG
  have hp' {l : ℕ} (h1 : l.Prime) (h2 : l ∣ Nat.card G) : (Nat.card G).minFac ≤ l :=
    minFac_le_of_dvd h1.two_le h2
```